crates/contract-client/src/client.rs

```diff
@@ -266,4 +292,40 @@ where
             })
             .collect())
     }
+
+    async fn all_gateways(&self, worker_id: U256) -> Result<Vec<GatewayCluster>, ClientError> {
+        lazy_static! {
+            pub static ref GATEWAYS_PAGE_SIZE: U256 = 10000.into();
+        }
+
+        let latest_block = self.l2_client.get_block_number().await?;
+
+        let mut clusters = HashMap::new();
+        for page in 0.. {
```

Comment:
I would do some fancy stream with `map().take_while().into_grouping_map().fold()` but a for loop is fine. :)